// jaeger-client/dist/src/index.js
'use strict';

var _configuration = require('./configuration');

var _configuration2 = _interopRequireDefault(_configuration);

var _configuration_env = require('./configuration_env');

var _configuration_env2 = _interopRequireDefault(_configuration_env);

var _span_context = require('./span_context');

var _span_context2 = _interopRequireDefault(_span_context);

var _span = require('./span');

var _span2 = _interopRequireDefault(_span);

var _tracer = require('./tracer');

var _tracer2 = _interopRequireDefault(_tracer);

var _const_sampler = require('./samplers/const_sampler');

var _const_sampler2 = _interopRequireDefault(_const_sampler);

var _probabilistic_sampler = require('./samplers/probabilistic_sampler');

var _probabilistic_sampler2 = _interopRequireDefault(_probabilistic_sampler);

var _rate_limiting_sampler = require('./samplers/rate_limiting_sampler');

var _rate_limiting_sampler2 = _interopRequireDefault(_rate_limiting_sampler);

var _remote_sampler = require('./samplers/remote_sampler');

var _remote_sampler2 = _interopRequireDefault(_remote_sampler);

var _priority_sampler = require('./samplers/experimental/priority_sampler');

var _priority_sampler2 = _interopRequireDefault(_priority_sampler);

var _tag_equals_sampler = require('./samplers/experimental/tag_equals_sampler');

var _tag_equals_sampler2 = _interopRequireDefault(_tag_equals_sampler);

var _composite_reporter = require('./reporters/composite_reporter');

var _composite_reporter2 = _interopRequireDefault(_composite_reporter);

var _in_memory_reporter = require('./reporters/in_memory_reporter');

var _in_memory_reporter2 = _interopRequireDefault(_in_memory_reporter);

var _logging_reporter = require('./reporters/logging_reporter');

var _logging_reporter2 = _interopRequireDefault(_logging_reporter);

var _noop_reporter = require('./reporters/noop_reporter');

var _noop_reporter2 = _interopRequireDefault(_noop_reporter);

var _remote_reporter = require('./reporters/remote_reporter');

var _remote_reporter2 = _interopRequireDefault(_remote_reporter);

var _text_map_codec = require('./propagators/text_map_codec');

var _text_map_codec2 = _interopRequireDefault(_text_map_codec);

var _zipkin_b3_text_map_codec = require('./propagators/zipkin_b3_text_map_codec');

var _zipkin_b3_text_map_codec2 = _interopRequireDefault(_zipkin_b3_text_map_codec);

var _test_util = require('./test_util');

var _test_util2 = _interopRequireDefault(_test_util);

var _tchannel_bridge = require('./tchannel_bridge');

var _tchannel_bridge2 = _interopRequireDefault(_tchannel_bridge);

var _prometheus = require('./metrics/prometheus');

var _prometheus2 = _interopRequireDefault(_prometheus);

var _opentracing = require('opentracing');

var opentracing = _interopRequireWildcard(_opentracing);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2016 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License. You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied. See the License for the specific language governing permissions and limitations under
// the License.

module.exports = {
  Configuration: _configuration2.default,
  initTracer: _configuration2.default.initTracer,
  initTracerFromEnv: _configuration_env2.default.initTracer,
  SpanContext: _span_context2.default,
  Span: _span2.default,
  Tracer: _tracer2.default,

  ConstSampler: _const_sampler2.default,
  ProbabilisticSampler: _probabilistic_sampler2.default,
  RateLimitingSampler: _rate_limiting_sampler2.default,
  RemoteSampler: _remote_sampler2.default,
  experimental: {
    PrioritySampler: _priority_sampler2.default,
    TagEqualsSampler: _tag_equals_sampler2.default
  },

  CompositeReporter: _composite_reporter2.default,
  InMemoryReporter: _in_memory_reporter2.default,
  LoggingReporter: _logging_reporter2.default,
  NoopReporter: _noop_reporter2.default,
  RemoteReporter: _remote_reporter2.default,

  TextMapCodec: _text_map_codec2.default,
  ZipkinB3TextMapCodec: _zipkin_b3_text_map_codec2.default,

  TestUtils: _test_util2.default,
  TChannelBridge: _tchannel_bridge2.default,
  PrometheusMetricsFactory: _prometheus2.default,
  opentracing: opentracing
};
